sessions/2019/january/session_21.md

```diff
@@ -0,0 +1,71 @@
+# SOFTWARE ENGINEERING IN JAVA
+
+## Session 21 (08/01/2019)
+
+- Corrections to fibonacci algorithm.
+- Implementing Logger and tests for it.
+
+### Notes
+
+Some improves where done to the code in fibonacci in order to make it easier to read.
+
+#### Logger
+
+In order to implement log capability we created the following interface:
+
+```java
+public interface Logger {
+    void log(String message);
+}
+```
+
+And after we implement that interface inside a DefaultLogger class:
+```java
+@Component
+public class DefaultLogger implements Logger { ... }
+```
+
+Notice that `@Component` Spring annotation was used in order we can use the Logger inside our project classes. Hence, for example in the `Calculator` class we are injecting our Logger interface:
+
+```java
+@Service
+public class Calculator {
+    private Logger logger;
+
+    Calculator (Logger logger) {
+        this.logger = logger;
+    }
+    ...
+}
+```
+
+If we not define our Logger implementation as `@Component`, Spring engine cannot know where that interface definition is.
+
+A thing to highlight is the fact that we are injecting the interface but not the particular class implementation which we define as `DefaultLogger`, however, Spring choose it automatically since its the only one that exists in our project for the `Logger` interface.
+
+**Note:** In tests we have another implementation for `Logger` interface but Spring takes it as another enviroment apart so don't recognize any interface definition for our project.
```

Comment:
I don't think the tests currently use spring at all. We just pass in the test logger to the class we're testing directly in the constructor.